app/experimenter/nimbus-ui/src/components/FormBranches/reducer.tsx

```diff
@@ -0,0 +1,162 @@
+import { getExperiment_experimentBySlug } from "../../types/getExperiment";
+
+export function formBranchesReducer(
+  state: FormBranchesState,
+  action: FormBranchesAction,
+): FormBranchesState {
+  const newState = { ...state };
+  switch (action.type) {
+    case "addBranch":
+      return addBranch(newState);
+    case "removeBranch":
+      return removeBranch(newState, action);
+    case "updateBranch":
+      return updateBranch(newState, action);
+    case "removeFeatureConfig":
+      return removeFeatureConfig(newState);
+    case "setFeatureConfig":
+      return setFeatureConfig(newState, action);
+    case "setEqualRatio":
+      return setEqualRatio(newState, action);
+    default:
+      return newState;
+  }
+}
+
+export const REFERENCE_BRANCH_IDX = -1;
+
+export type FormBranchesState = {
+  equalRatio: boolean;
+  featureConfig: getExperiment_experimentBySlug["featureConfig"];
+  referenceBranch: getExperiment_experimentBySlug["referenceBranch"];
+  treatmentBranches: getExperiment_experimentBySlug["treatmentBranches"];
+};
+
+export type FormBranchesAction =
+  | AddBranchAction
+  | RemoveBranchAction
+  | UpdateBranchAction
+  | RemoveFeatureConfigAction
+  | SetEqualRatioAction
+  | SetFeatureConfigAction;
+
+type AddBranchAction = {
+  type: "addBranch";
+};
+
+function addBranch(newState: FormBranchesState) {
+  // TODO: figure out some default values to stick here
+  const newBranch = {
+    name: "",
+    // TODO: maybe need to slugify client-side?
```

Comment:
Do we need issues for this and the above TODO?

Reply:
Planning to either resolve this as part of this PR if it's behind the add-branch issues - or, it can be a part of EXP-505, because I'm expecting some of this will need refinement when joined up with a mutation